diff --git a/drivers/extcon/extcon-usbc-virtual-pd.c b/drivers/extcon/extcon-usbc-virtual-pd.c
index e0713670e..33b0db940 100644
--- a/drivers/extcon/extcon-usbc-virtual-pd.c
+++ b/drivers/extcon/extcon-usbc-virtual-pd.c
@@ -146,7 +146,7 @@ static int vpd_extcon_parse_dts(struct vpd_extcon *vpd)
 
 	vpd->data_role = vpd_extcon_data_role(vpd);
 
-	vpd->det_gpio = devm_gpiod_get_optional(dev, "det", GPIOD_OUT_LOW);
+	vpd->det_gpio = devm_gpiod_get_optional(dev, "det", GPIOD_IN);
 	if (IS_ERR(vpd->det_gpio)) {
 		ret = PTR_ERR(vpd->det_gpio);
 		dev_warn(dev, "failed to get det gpio: %d\n", ret);
